internal/container/container.go

```diff
@@ -414,8 +440,145 @@ func truncateString(s string, maxLength int) string {
 	return s
 }
 
-// Cleanup all docker containers for the given container
-func (c *Container) CleanupArtifacts() error {
+// CleanupArtifacts will clean up all docker artifacts for the given container
+func (c *Container) CleanupArtifacts(keepLatest bool) error {
 	slog.Info("cleanup artifacts", "name", c.Name)
-	return c.dockerClient.CleanupArtifactsForLabel("io.flowpipe.name", c.Name)
+	return c.dockerClient.CleanupArtifactsForLabel("io.flowpipe.name", c.Name, docker.WithSkipLatest(keepLatest))
+}
+
+func (c *Container) IsFromSource() bool {
+	return c.Image == "" && c.Source != ""
+}
+
+func (c *Container) Watch() error {
+	if c.Image != "" {
+		return nil
+	}
+
+	wd := viper.GetString(constants.ArgModLocation)
+	df := filepath.Join(wd, c.Source)
+
+	c.watcher = watcher.New()
+	c.watcher.SetMaxEvents(1)
+	if err := c.watcher.Add(df); err != nil {
+		return perr.BadRequestWithMessage("failed to add watch for container source: " + c.Source)
+	}
+
+	// watch for changes
+	go func() {
+		for {
+			select {
+			case event := <-c.watcher.Event:
+				go func() {
+					slog.Info("container watch event", "event", event)
+					if err := c.Build(); err != nil {
+						slog.Error(fmt.Sprintf("failed to build container image %s, got error: %v", c.Name, err), "error", err, "containerName", c.Name)
+					}
+				}()
+			case err := <-c.watcher.Error:
+				slog.Error("file watcher error", "error", err)
+			case <-c.watcher.Closed:
+				return
+			}
+		}
+	}()
+
+	// watcher in background
+	go func() {
+		if err := c.watcher.Start(time.Millisecond * 1000); err != nil {
+			slog.Error("failed to start file watcher", "container", c.Name, "error", err)
+		}
+	}()
+
+	return nil
+}
+
+func (c *Container) Build() error {
+	err := c.build()
+	if err != nil {
+		return err
+	}
+
+	// If a build was queued, run it now.
+	if c.buildQueued {
+		c.buildQueued = false
+		return c.build()
```

Comment:
this should be c.Build not c.build.

I suggest calling it Build and BuildOne just like how we do it in the function step. If you follow this logic it won't work.

The build won't be queued, instead it will just fail after the 2nd retry.